feature-libs/product-configurator/rulebased/root/interceptor/cpq-configurator-rest.interceptor.ts

```diff
@@ -0,0 +1,104 @@
+import {
+  HttpErrorResponse,
+  HttpEvent,
+  HttpHandler,
+  HttpInterceptor,
+  HttpRequest,
+  HttpResponse,
+} from '@angular/common/http';
+import { Injectable } from '@angular/core';
+import { Observable, throwError } from 'rxjs';
+import { catchError, switchMap, take, tap } from 'rxjs/operators';
+
+import { CpqAccessData } from './cpq-access-data.models';
+import { CpqAccessStorageService } from './cpq-access-storage.service';
+
+const HEADER_ATTR_CPQ_SESSION_ID = 'x-cpq-session-id';
+const HEADER_ATTR_CPQ_NO_COOKIES = 'x-cpq-disable-cookies';
+
+export const CPQ_CONFIGURATOR_VIRTUAL_ENDPOINT =
+  'cpq-configurator-virtual-enpoint';
+
+@Injectable({
+  providedIn: 'root',
+})
+export class CpqConfiguratorRestInterceptor implements HttpInterceptor {
+  constructor(protected cpqAccessStorageService: CpqAccessStorageService) {}
+
+  intercept(
+    request: HttpRequest<any>,
+    next: HttpHandler
+  ): Observable<HttpEvent<any>> {
+    if (!request.url.startsWith(CPQ_CONFIGURATOR_VIRTUAL_ENDPOINT)) {
+      return next.handle(request);
+    }
+    return this.cpqAccessStorageService.getCachedCpqAccessData().pipe(
+      take(1), // avoid request being re-executed when token expires
+      switchMap((cpqData) => {
+        return next.handle(this.enrichHeaders(request, cpqData)).pipe(
+          catchError((errorResponse: any) => {
+            return this.handleError(errorResponse, next, request);
+          }),
+          tap((response) => this.extractCpqSessionId(response, cpqData))
```

Comment:
I'm having hard time understanding why the Interceptor has to _mutate_ the object `cpqData`. That object is emitted by a dependency Service. Ideally the service should extract `cpqSessionId` in itself, becasue it's already the "producer" of the value. Alternatively... the interceptor could call some setter method on the service. Nevertheless IMHO when consuming emitted values of an observable we should never mutate it - but instead push a fresh object to the source observable. Otherwise it would be unpredicatable silent change for any other consumers of that observable.

Reply:
pick

Reply:
@Uli-Tiger Only now I get that it makes sense to store `cpqSessionId` in the interceptor instead of auth storage service.